File: test/CSharp/CodeCracker.Test/Usage/ArgumentExceptionTests.cs
using CodeCracker.CSharp.Usage;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using System.Threading.Tasks;
using Xunit;

namespace CodeCracker.Test.CSharp.Usage
{
    public class ArgumentExceptionTests : CodeFixVerifier<ArgumentExceptionAnalyzer, ArgumentExceptionCodeFixProvider>
    {
        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentAnalyzerCreatesDiagnostic()
        {
            var test = Wrap(@"
            public async Task Foo(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var expected = new DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(11, 56)
                .WithMessage("Type argument 'c' is not in the argument list.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionInCtorWithInvalidArgumentAnalyzerCreatesDiagnostic()
        {
            var test = Wrap(@"
            public TypeName(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var expected = new DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(11, 56)
                .WithMessage("Type argument 'c' is not in the argument list.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentAndApplyingFirstFixUsesFirstParameter()
        {
            var test = Wrap(@"
            public async Task Foo(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var fixtest = Wrap(@"
            public async Task Foo(int a, int b)
            {
                throw new ArgumentException(""message"", ""a"");
            }");
            await VerifyCSharpFixAsync(test, fixtest, 0);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentAndApplyingSecondFixUsesSecondParameter()
        {
            var test = Wrap(@"
            public async Task Foo(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var fixtest = Wrap(@"
            public async Task Foo(int a, int b)
            {
                throw new ArgumentException(""message"", ""b"");
            }");
            await VerifyCSharpFixAsync(test, fixtest, 1);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInCtorAndApplyingFirstFixUsesFirstParameter()
        {
            var test = Wrap(@"
            public TypeName(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var fixtest = Wrap(@"
            public TypeName(int a, int b)
            {
                throw new ArgumentException(""message"", ""a"");
            }");
            await VerifyCSharpFixAsync(test, fixtest, 0);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInCtorAndApplyingSecondFixUsesSecondParameter()
        {
            var test = Wrap(@"
            public TypeName(int a, int b)
            {
                throw new ArgumentException(""message"", ""c"");
            }");

            var fixtest = Wrap(@"
            public TypeName(int a, int b)
            {
                throw new ArgumentException(""message"", ""b"");
            }");
            await VerifyCSharpFixAsync(test, fixtest, 1);
        }

        [Fact]
        public async Task IgnoresArgumentExceptionObjectsInFields()
        {
            var test = Wrap(@"
            ArgumentException ex = new ArgumentException(""message"", ""paramName"");
            ");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IgnoresArgumentExceptionObjectsInGetAccessors()
        {
            var test = Wrap(@"
            public string StrangePropertyThatThrowsArgumentExceptionInsideGet
            { get { throw ArgumentException(""message"", ""paramName""); } }
            ");
            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionInSetPropertyArgumentNameShouldBeValue()
        {
            var test = Wrap(@"
            public string RejectsEverythingProperty
            {
                get { return null; } 
                set { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var expected = new DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(12, 62)
                .WithMessage("Type argument 'c' is not in the argument list.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInSetAccessorAndApplyingFixUsesParameter()
        {
            var test = Wrap(@"
            public string RejectsEverythingProperty
            {
                get { return null; } 
                set { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var fixtest = Wrap(@"
            public string RejectsEverythingProperty
            {
                get { return null; } 
                set { throw new ArgumentException(""message"", ""value""); } 
            }
            ");
            await VerifyCSharpFixAsync(test, fixtest);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionInGetPropertyWithIndexersArgumentNameShouldBeInParameterList()
        {
            var test = Wrap(@"
            public string this[int a, int b]
            {
                get { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var expected = new DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(11, 62)
                .WithMessage("Type argument 'c' is not in the argument list.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInIndexerGetAccessorAndApplyingFirstFixUsesFirstParameter()
        {
            var test = Wrap(@"
            public string this[int a, int b]
            {
                get { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var fixtest = Wrap(@"
            public string this[int a, int b]
            {
                get { throw new ArgumentException(""message"", ""a""); } 
            }
            ");
            await VerifyCSharpFixAsync(test, fixtest, 0);
        }


        [Fact]
        public async Task WhenThrowingArgumentExceptionInLambdaArgumentNameShouldBeInParameterList()
        {
            var test = Wrap(@"
            Action<int> action = (p) => { throw new ArgumentException(""message"", ""paramName""); };
            ");

            var expected = new DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), DiagnosticSeverity.Warning)
                .WithLocation(9, 82)
                .WithMessage("Type argument 'paramName' is not in the argument list.");

            await VerifyCSharpDiagnosticAsync(test, expected);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInLambdasAndApplyingFixUsesLambdaParameter()
        {
            var test = Wrap(@"
            Action<int> action = (p) => { throw new ArgumentException(""message"", ""paramName""); };
            ");

            var fixtest = Wrap(@"
            Action<int> action = (p) => { throw new ArgumentException(""message"", ""p""); };
            ");
            await VerifyCSharpFixAsync(test, fixtest);
        }


        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInSimpleLambdasAndApplyingFixUsesLambdaParameter()
        {
            var test = Wrap(@"
            Action<int> action = p => { throw new ArgumentException(""message"", ""paramName""); };

            void Foo(string a)
            {
                Action<int> action = param => { throw new ArgumentException(""message"", ""paramName""); };
            } 
            ");

            var fixtest = Wrap(@"
            Action<int> action = p => { throw new ArgumentException(""message"", ""p""); };

            void Foo(string a)
            {
                Action<int> action = param => { throw new ArgumentException(""message"", ""param""); };
            } 
            ");
            await VerifyCSharpFixAsync(test, fixtest);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInIndexerGetAccessorAndApplyingSecondFixUsesSecondParameter()
        {
            var test = Wrap(@"
            public string this[int a, int b]
            {
                get { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var fixtest = Wrap(@"
            public string this[int a, int b]
            {
                get { throw new ArgumentException(""message"", ""b""); } 
            }
            ");
            await VerifyCSharpFixAsync(test, fixtest, 1);
        }

        [Fact]
        public async Task WhenThrowingArgumentExceptionWithInvalidArgumentInIndexerSetAccessorAndApplyingThirdFixUsesValue()
        {
            var test = Wrap(@"
            public string this[int a, int b]
            {
                get { return null; }
                set { throw new ArgumentException(""message"", ""c""); } 
            }
            ");

            var fixtest = Wrap(@"
            public string this[int a, int b]
            {
                get { return null; }
                set { throw new ArgumentException(""message"", ""value""); } 
            }
            ");
            await VerifyCSharpFixAsync(test, fixtest, 2);
        }

        [Fact]
        public async Task IgnoresArgumentExceptionObjectsInSetAccessorsOfIndexersThatUsesValue()
        {
            var test = Wrap(@"
            public string this[int a, int b]
            {
                get { return null; }
                set { throw new ArgumentException(""message"", ""value""); } 
            }
            ");

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        [Fact]
        public async Task IgnoresArgumentExceptionObjectsInInitializerOfAutoProperties()
        {
            var test = Wrap(@"
            ArgumentException Exception { get; } = new ArgumentException(""message"", ""paramName"");
            ");

            await VerifyCSharpHasNoDiagnosticsAsync(test);
        }

        static string Wrap(string code)
        {
            return @"
    using System;

    namespace ConsoleApplication1
    {
        class TypeName
        {
            " + code + @"
        }
    }";

        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Usage/ArgumentExceptionTests.vb
Imports CodeCracker.VisualBasic.Usage
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Usage
    Public Class ArgumentExceptionTests
        Inherits CodeFixVerifier(Of ArgumentExceptionAnalyzer, ArgumentExceptionCodeFixProvider)

        Shared Function Wrap(code As String) As String
            Return "
Using System

Namespace ConsoleApplication1
    Class TypeName
        " & code & "
    End Class
End Namespace"
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionWithInvalidArgumentAnalyzerCreatesDiagnostic() As Task
            Dim test = Wrap("
Public Async Function Foo(a As Integer, b As Integer) As Task
    Throw New ArgumentException(""message"", ""c"")
End Function
")

            Dim expected = New DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(8, 44) _
                .WithMessage("Type argument 'c' is not in the argument list.")

            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionInConstrucorWithInvalidARgumentAnalyzerCreatesDiagnostic() As Task
            Dim test = Wrap("
Public Sub New(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""c"")
End Sub
")

            Dim expected = New DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(8, 44) _
                .WithMessage("Type argument 'c' is not in the argument list.")

            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionWithInvalidArgumentAndApplyingFirstFixUsesFirstParameter() As Task
            Dim test = Wrap("
Public Sub Foo(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""c"")
End Sub
")

            Dim fix = Wrap("
Public Sub Foo(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""a"")
End Sub
")
            Await VerifyBasicFixAsync(test, fix, 0)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionWithInvalidArgumentAndApplyingSecondFixUsesSecondParameter() As Task
            Dim test = Wrap("
Public Sub Foo(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""c"")
End Sub
")

            Dim fix = Wrap("
Public Sub Foo(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""b"")
End Sub
")
            Await VerifyBasicFixAsync(test, fix, 1)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionWithInvalidArgumentInConstructorAndApplyingFirstFixUsesFirstParameter() As Task
            Dim test = Wrap("
Public Sub New(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""c"")
End Sub
")

            Dim fix = Wrap("
Public Sub New(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""a"")
End Sub
")
            Await VerifyBasicFixAsync(test, fix, 0)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionWithInvalidArgumentInConstructorAndApplyingFirstFixUsesSecondParameter() As Task
            Dim test = Wrap("
Public Sub New(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""c"")
End Sub
")

            Dim fix = Wrap("
Public Sub New(a As Integer, b As Integer)
    Throw New ArgumentException(""message"", ""b"")
End Sub
")
            Await VerifyBasicFixAsync(test, fix, 1)
        End Function

        <Fact>
        Public Async Function IgnoresAgrumentExceptionObjectsInFields() As Task
            Const test = "Dim ex As New ArgumentException(""message"", ""paramName"")"
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function IgnoresArgumentExceptionObjectsInGetAccessors() As Task
            Dim test = Wrap("
        Public Property NewProperty() As String
            Get
                Throw New ArgumentException(""message"", ""paramName"")
            End Get
            Set(ByVal value As String)
            End Set
        End Property
")
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionInSetPropertyArgumentNameShouldBeValue() As Task
            Dim test = Wrap("
        Public Property NewProperty() As String
            Get
            End Get
            Set(ByVal value As String)
                Throw New ArgumentException(""message"", ""paramName"")
            End Set
        End Property
")
            Dim expected = New DiagnosticResult(DiagnosticId.ArgumentException.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Warning) _
                .WithLocation(11, 56) _
                .WithMessage("Type argument 'paramName' is not in the argument list.")
            Await VerifyBasicDiagnosticAsync(test, expected)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgmentExceptionWithInvalidArgumentInSetAccessorAndApplyingFixUsesParameter() As Task
            Dim test = Wrap("
        Public Property NewProperty() As String
            Get
                Return Nothing
            End Get
            Set(ByVal value As String)
                Throw New ArgumentException(""message"", ""paramName"")
            End Set
        End Property
")
            Dim fix = Wrap("
        Public Property NewProperty() As String
            Get
                Return Nothing
            End Get
            Set(ByVal value As String)
                Throw New ArgumentException(""message"", ""value"")
            End Set
        End Property
")

            Await VerifyBasicFixAsync(test, fix)
        End Function

        <Fact>
        Public Async Function WhenThrowingArgumentExceptionInMultilineLambdaAndApplyingFix() As Task
            Dim test = Wrap("
        Dim a = Function(p) 
                   Throw New ArgumentException(""message"", ""paramName"")
                End Function
        Sub Foo(a As String)
            Dim action = Sub(p1)
                Throw New ArgumentException(""message"", ""paramName"")
            End Sub        
        End Sub
")
            Dim fix = Wrap("
        Dim a = Function(p) 
                   Throw New ArgumentException(""message"", ""p"")
                End Function
        Sub Foo(a As String)
            Dim action = Sub(p1)
                Throw New ArgumentException(""message"", ""p1"")
            End Sub        
        End Sub
")

            Await VerifyBasicFixAsync(test, fix)
        End Function

        <Fact>
        Public Async Function IgnoresArgumentExceptionObjectsInInitializerOfAutoProperties() As Task
            Const test = "
ReadOnly Property Test = New ArgumentException(""message"", ""paramName"")
"
            Await VerifyBasicHasNoDiagnosticsAsync(test)
        End Function

    End Class
End Namespace